Need tests to ensure edit-and-continue works with expression-bodied methods

ivanbasov opened the pull request:
Fixes #17681

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/LineEditTests.cs:
```diff
@@ -1417,6 +1417,50 @@ class C
 
+        [Fact]
+        public void Event_ExpressionBodyToBlock()
+        {
```
These tests shouldn't be in "LineEditTests" class. The tests here validate that when the body is not changed, just moved to a different line we calculate line deltas correctly.

tmat replied:
I think "TopLevelEditingTests" is the right place for them.

---
In reply to: [128395722](https://github.com/dotnet/roslyn/pull/20931#discussion_r128395722) [](ancestors = 128395722)

ivanbasov replied:
ok. thanks!

---
In reply to: [128395783](https://github.com/dotnet/roslyn/pull/20931#discussion_r128395783) [](ancestors = 128395783,128395722)

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:
```diff
@@ -4393,6 +4457,114 @@ partial class C
                 Diagnostic(ErrorCode.ERR_MemberNameSameAsType, "C").WithArguments("C").WithLocation(4, 18));
         }
 
+        [Fact]
```
>Fact [](start = 9, length = 4)

Let's add a link to the github issue using WorkItem() attribute, so one can find more context.

ivanbasov replied:
Sure. Thank you!

---
In reply to: [128396148](https://github.com/dotnet/roslyn/pull/20931#discussion_r128396148) [](ancestors = 128396148)

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:
```diff
@@ -4393,6 +4457,114 @@ partial class C
 
+        [Fact]
+        public void Constructor_BlockToExpressionBody()
+        {
```
>{ [](start = 8, length = 1)

I'd also add test of constructors with initializers:

public D(int value) : base(value) => ...
public D(int value) : base(value) { ... }

ivanbasov replied:
added. thank you!

---
In reply to: [128396271](https://github.com/dotnet/roslyn/pull/20931#discussion_r128396271) [](ancestors = 128396271)

tmat commented:
:clock1:

tmat commented:
>         }

I think we are missing the case of property accessor switch to expr body:

int P { get { return 1; } }
int P { get => 1; }

Now, the setter wouldn't work since it doesn't return anything. I think that's fine, let's have a test for that though.



---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:6517 in 87db972. [](commit_id = 87db972aa658f62618c35f2e6d0102074396bba5, deletion_comment = False)

tmat commented:
>         }

Also for indexer...


---
In reply to: [316556365](https://github.com/dotnet/roslyn/pull/20931#issuecomment-316556365) [](ancestors = 316556365)

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:6517 in 87db972. [](commit_id = 87db972aa658f62618c35f2e6d0102074396bba5, deletion_comment = False)

ivanbasov commented:
>         }

I hope we have tests for indexers: [http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/EditAndContinue/ActiveStatementTests.Methods.cs,555](http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/EditAndContinue/ActiveStatementTests.Methods.cs,555) and below

---
In reply to: [316557222](https://github.com/dotnet/roslyn/pull/20931#issuecomment-316557222) [](ancestors = 316557222,316556365)

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:6517 in 87db972. [](commit_id = 87db972aa658f62618c35f2e6d0102074396bba5, deletion_comment = False)

tmat commented:
>         }

Just tried this. Doesn't work right now, so there is work to do.

---
In reply to: [316557222](https://github.com/dotnet/roslyn/pull/20931#issuecomment-316557222) [](ancestors = 316557222,316556365)

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:6517 in 87db972. [](commit_id = 87db972aa658f62618c35f2e6d0102074396bba5, deletion_comment = False)

ivanbasov commented:
>         }

We also have tests for properties: [http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/EditAndContinue/ActiveStatementTests.Methods.cs,441](http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/EditAndContinue/ActiveStatementTests.Methods.cs,441)

---
In reply to: [316556365](https://github.com/dotnet/roslyn/pull/20931#issuecomment-316556365) [](ancestors = 316556365)

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:6517 in 87db972. [](commit_id = 87db972aa658f62618c35f2e6d0102074396bba5, deletion_comment = False)

ivanbasov commented:
>         }

Fixed for ctors, descrutrors, indexers and events. Added unit tests.

---
In reply to: [316559640](https://github.com/dotnet/roslyn/pull/20931#issuecomment-316559640) [](ancestors = 316559640,316556365)

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:6517 in 87db972. [](commit_id = 87db972aa658f62618c35f2e6d0102074396bba5, deletion_comment = False)

tmat commented on src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs:
```diff
@@ -187,7 +187,8 @@ public static SyntaxNode TryGetEffectiveGetterBody(ArrowExpressionClauseSyntax p
             }
 
-            return accessorList?.Accessors.Where(a => a.IsKind(SyntaxKind.GetAccessorDeclaration)).FirstOrDefault()?.Body;
+            var firstAccessor = accessorList?.Accessors.Where(a => a.IsKind(SyntaxKind.GetAccessorDeclaration)).FirstOrDefault();
```
>firstAccessor [](start = 16, length = 13)

i'd call it firstGetter as we are filtering out other accessors

ivanbasov replied:
ok. thanks!

---
In reply to: [128831749](https://github.com/dotnet/roslyn/pull/20931#discussion_r128831749) [](ancestors = 128831749)

tmat commented on src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs:
```diff
@@ -187,7 +187,8 @@ public static SyntaxNode TryGetEffectiveGetterBody(ArrowExpressionClauseSyntax p
 
-            return accessorList?.Accessors.Where(a => a.IsKind(SyntaxKind.GetAccessorDeclaration)).FirstOrDefault()?.Body;
+            var firstAccessor = accessorList?.Accessors.Where(a => a.IsKind(SyntaxKind.GetAccessorDeclaration)).FirstOrDefault();
+            return (SyntaxNode)firstAccessor?.Body ?? firstAccessor?.ExpressionBody?.Expression;
```
>?.Body ?? firstAccessor?.E [](start = 44, length = 26)

A lot of ? here. Perhaps change to 
if (firstGetter == null)
{
  return null;
}

to make it clearer

ivanbasov replied:
ok. thanks!

---
In reply to: [128832056](https://github.com/dotnet/roslyn/pull/20931#discussion_r128832056) [](ancestors = 128832056)

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:
```diff
@@ -6344,6 +6586,108 @@ public void Property_BlockBodyToExpressionBody2()
+            edits.VerifySemantics(ActiveStatementsDescription.Empty, new[]
+            {
+                SemanticEdit(SemanticEditKind.Update, c => c.GetMember("C.get_P"), preserveLocalVariables: false),
+                SemanticEdit(SemanticEditKind.Update, c => c.GetMember("C.set_P"), preserveLocalVariables: false)
```
>SemanticEdit(SemanticEditKind.Update, c => c.GetMember("C.set_P"), preserveLocalVariables: false) [](start = 16, length = 97)

I guess there is an update because the column changed.
Can we change the test to avoid that? Just add some spaces before "set" keyword in src1 to make sure the position of the setter doesn't change.

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:
```diff
@@ -6737,6 +7081,180 @@ public void IndexerWithExpressionBody_Update()
+            edits.VerifySemantics(ActiveStatementsDescription.Empty, new[]
+            {
+                SemanticEdit(SemanticEditKind.Update, c => c.GetMember("C.get_Item"), preserveLocalVariables: false),
+                SemanticEdit(SemanticEditKind.Update, c => c.GetMember("C.set_Item"), preserveLocalVariables: false)
```
>C.set_Item"), preserv [](start = 72, length = 21)

The same as above.

ivanbasov replied:
ok. let us try

---
In reply to: [128833161](https://github.com/dotnet/roslyn/pull/20931#discussion_r128833161) [](ancestors = 128833161)

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:
```diff
@@ -7183,6 +7701,54 @@ class C
+using System;
+public class C
+{
+   event Action E { add { F(); } remove => F(); }
```
>=> F();  [](start = 40, length = 8)

Perhaps change this also to { }, so that we cover remove handler changing from block to expression body as well.

ivanbasov replied:
ok and below

---
In reply to: [128833458](https://github.com/dotnet/roslyn/pull/20931#discussion_r128833458) [](ancestors = 128833458)

agocke commented on src/Features/CSharp/Portable/EditAndContinue/SyntaxUtilities.cs:
```diff
@@ -187,7 +187,13 @@ public static SyntaxNode TryGetEffectiveGetterBody(ArrowExpressionClauseSyntax p
+                return null;
+            }
+
+            return (SyntaxNode)firstGetter.Body ?? firstGetter.ExpressionBody?.Expression;
```
It looks like we're checking in a few places for a non-`null` `ExpressionBody` with a `null` expression. Is this actually possible? If so, I think there should be a test for this.

tmat replied:
It's possible if the property is abstract, in an interface or in presence of syntax errors. I don't think we have EnC scenarios where it can happen. If we had a unit test targeting this specific helper then I'd say yes, we should have a test case for this. We don't have unit tests for all the helpers currently.

tmat commented:
:shipit:

tmat commented:
@jaredpar This is IDE change, not a compiler change.

jaredpar commented:
Okay. can we remove roslyn-compiler as a requestned reviewer?
